### PR TITLE
Navigate up and down. 

### PR DESCRIPTION
**Pull request by sriramgd**
For https://github.com/OmniSharp/omnisharp-roslyn/issues/97

### CODE REVIEW

**Comment by sriramgd** [comment]
This is ready for review.

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                {
+                    break;
+                }
+                stack.AddRange(node.ChildNodes);
+            }
+            return new ObjectResult(ret);
+        }
+
+        [HttpPost("navigateup")]
+        public async Task<IActionResult> NavigateUp(Request request)
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: put the Controller actions above the private methods.... it's a convention we have

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                qf.Column = node.Location.Column;
+                retVal = true;
+            }
+            return retVal;
+        };
+
+        private async Task<IActionResult> Navigate(Request request, bool toBreak, Func<Request, FileMemberElement, GotoDefinitionResponse, bool> navigateFunc)
+        {
+            var stack = new List<FileMemberElement>(await StructureComputer.Compute(_workspace.GetDocuments(request.FileName)));
+            var ret = new GotoDefinitionResponse();
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Shouldn't really be returning a GotoDefinitionResponse here (although we need an object with the same signature)

Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: ret is a bad name.

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                stack.Remove(node);
+                //Need to call the passed in up or down delegate
+                var found = navigateFunc(request, node, ret);
+                if (toBreak && found)
+                {
+                    break;
+                }
+                stack.AddRange(node.ChildNodes);
+            }
+            return new ObjectResult(ret);
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Just return the object instead of the ObjectResult. It will make it simpler to test and we need POCOs for the stdout interface I believe.

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+using OmniSharp.Models;
+using System;
+
+namespace OmniSharp
+{
+    public partial class OmnisharpController
+    {
+        private Func<Request, FileMemberElement, GotoDefinitionResponse, bool> NavigateUpFunc = (request, node, qf) =>
+        {
+            bool retVal = false;
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: retVal is a bad name

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            if (node.Location.Line < request.Line || (node.Location.Line == request.Line && node.Location.Column < request.Column))
+            {
+                qf.Line = node.Location.Line;
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: what's qf?

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            return await Navigate(request, toBreak, NavigateUpFunc);
+        }
+
+        [HttpPost("navigatedown")]
+        public async Task<IActionResult> NavigateDown(Request request)
+        {
+            bool toBreak = true;
+            return await Navigate(request, toBreak, NavigateDownFunc);
+        }
+
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: delete this line

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc;
+using OmniSharp.Models;
+using System;
```
Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: `using System` should be first

**Comment by sriramgd** [comment]
Updated based on code review. Refactored not to depend on any implicit ordering of nodes. Now both navigate up and down are symmetric and work the same way.
Tested with properties, methods and classes.

**Comment by davidfowl** [comment]
Where are the unit tests?

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                response.Column = closestNode.Location.Column;
+            }
+            else if (thisNode != null)
+            {
+                response.Column = thisNode.Location.Column;
+            }
+            return response;
+        }
+
+        private Func<FileMemberElement, FileMemberElement, Request, bool> IsCloserNodeUpFunc = (candidateClosestNode, closestNode, request) =>
```
Comment (davidfowl) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Why re these funcs instead of methods?

File: src/OmniSharp/Models/NavigateResponse.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+namespace OmniSharp.Models
+{
+    public class NavigateResponse
+    {
+        public string FileName { get; set; }
```
Comment (jrieken) [inline src/OmniSharp/Models/NavigateResponse.cs]: will you ever end up in a different file? Or should this be a a _code pointer_ thing (line and position)? Iff so it should have a different name

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                if (node.Location.Line == request.Line)
+                {
+                    thisNode = node;
+                }
+                stack.AddRange(node.ChildNodes);
+            }
+
+            //If there is a closest node, use its line and column.
+            //or if we are on the last node, adjust column.
+            //if we are above the first or below the first node, do nothing.
```
Comment (jrieken) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: how about (optionally, defined by request) starting at the other end and you are the top/bottom most node?

**Comment by Yantrio** [comment]
Do we want this for all filemembers, or just methods/properties ?

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs (hunk 8 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Added unit tests - NavigateUpDownFacts.cs.
I could use methods, but I would then need either a delegate definition to pass the methods in as parameters to the Navigate method or an if-else statement in the Navigate method.

File: src/OmniSharp/Models/NavigateResponse.cs (hunk 9 above)
Comment (sriramgd) [inline src/OmniSharp/Models/NavigateResponse.cs]: This will not end up in a different file. Does not try follow a partial, since there is no real order for the partial. Is there any use case you had in mind?

**Comment by sriramgd** [comment]
Updated based on second review. Unit tests added. Ready for another review.

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs (hunk 10 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Since this will be additive and not a breaking update, could we do this as a separate PR?

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs (hunk 8 above)
Comment (davidfowl) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: You can pass a method group. You don't have to define them as Func<...> to pass them as functions. The compiler does that for you. Turn these into methods.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 11]
```diff
@@ -0,0 +1,845 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using OmniSharp.Filters;
+using OmniSharp.Models;
+using Xunit;
+
+namespace OmniSharp.Tests
+{
+    enum NavigateDirection {UP, DOWN}
```
Comment (davidfowl) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: This looks weird. Can you expand this so that it looks like:

```
enum NavigateDirection 
{
    UP, 
    DOWN
}
```

File: src/OmniSharp/Models/NavigateResponse.cs (hunk 9 above)
Comment (jrieken) [inline src/OmniSharp/Models/NavigateResponse.cs]: Maybe not have the FileName than?

**Comment by nosami** [comment]
Nitpick - these tests would be clearer if we had another symbol in the source (similar to $) to show where the cursor will finish up.

**Comment by nosami** [comment]
Also, am I correct in thinking that the cursor finishes up on the method name, not the first character of the line? Is that what we want?

**Comment by nosami** [comment]
I added some keybindings for this over at https://github.com/nosami/omnisharp-demo/commit/9f374dbd1658e63d6071fb3d407a927fa18f1a72

Works well!

File: src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs (hunk 8 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs]: Done. Didnt know that before, thanks.

File: src/OmniSharp/Models/NavigateResponse.cs (hunk 9 above)
Comment (sriramgd) [inline src/OmniSharp/Models/NavigateResponse.cs]: Good point, fixed.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 11 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Done.

**Comment by sriramgd** [comment]
Tests refactored to use a source and target marker. I prefer ending up on the definition of the class/method/property/field. Maybe the client can move to start of line etc. as required? Comments squashed in git (I think).

Ready for third review.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 12]
```diff
@@ -0,0 +1,751 @@
...
+using System.Threading.Tasks;
+using OmniSharp.Filters;
+using OmniSharp.Models;
+using Xunit;
+
+namespace OmniSharp.Tests
+{
+    enum NavigateDirection 
+    {
+      UP,
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Indentation is off.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 13]
```diff
@@ -0,0 +1,751 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using OmniSharp.Filters;
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: System.Collections.Generic, System.Linq and OmniSharp.Filters aren't used.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 14]
```diff
@@ -0,0 +1,751 @@
...
+    enum NavigateDirection 
+    {
+      UP,
+      DOWN
+    }
+
+    public class NavigateUpDownFacts
+    {
+
+        private string foo = "test";
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: what's this? :) delete the line above and below too

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 15]
```diff
@@ -0,0 +1,751 @@
...
+            {
+                return await controller.NavigateUp(request);
+            }
+            else 
+            {
+                return await controller.NavigateDown(request);
+            }
+        }
+
+        private async Task AssertPosition(string fileContent, NavigateDirection navigateDirection, int finalCursorColumnOffset = 1)
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: I don't see why this is needed. 

Should always be 0 or some constant. It shouldn't ever change. Looks as if it's masking some bug in the test code, because I can see that this code works.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 16]
```diff
@@ -0,0 +1,751 @@
...
+      UP,
+      DOWN
+    }
+
+    public class NavigateUpDownFacts
+    {
+
+        private string foo = "test";
+
+        private async Task<NavigateResponse> SendRequest(OmnisharpWorkspace workspace, string fileName, string fileContent, NavigateDirection upOrDown)
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: not critical, but I'd prefer it if test helper methods came after the tests.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs [hunk 17]
```diff
@@ -0,0 +1,751 @@
...
+            using Microsoft.AspNet.Mvc;
+            using OmniSharp.Models;
+
+            namespace OmniSharp
+            {
+                public class NavigateController
+                {
+                    private string text;
+
+                    public string M%oreText
```
Comment (nosami) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Why are some of the these percent markers one character in?

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 12 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Done.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 13 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Done, sorry missed that.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 14 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Was testing property navigation using this file :) Removed.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 15 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Right. Now I realize I went adjusting for a bug instead of fixing the cause. Fixed.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 17 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Was a bug in the test code. Fixed.

File: tests/OmniSharp.Tests/NavigateUpDownFacts.cs (hunk 16 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/NavigateUpDownFacts.cs]: Done.

**Comment by sriramgd** [comment]
Ready for fourth :) review.

**Comment by nosami** [comment]
Looks good to me. Thanks for this :)
